src/blocks/post-carousel/post-item.js

```diff
@@ -0,0 +1,106 @@
+import {
+	Disabled,
+} from '@wordpress/components';
+import { PlainText } from '@wordpress/block-editor';
+import { RawHTML } from '@wordpress/element';
+import { withSelect } from '@wordpress/data';
+// Disable reason: We choose to use unsafe APIs in our codebase.
+// eslint-disable-next-line @wordpress/no-unsafe-wp-apis
+import { __experimentalGetSettings, dateI18n, format } from '@wordpress/date';
+
+const PostItem = ( {
+	post,
+	setAttributes,
+	attributes,
+} ) => {
+	const {
+		excerptLength,
+		displayPostLink,
+		displayPostDate,
+		displayPostContent,
+		postLink,
+	} = attributes;
+
+	const featuredImageUrl = post.featured_media_object ? post.featured_media_object.source_url : null;
+	const featuredImageStyle = 'url(' + featuredImageUrl + ')';
+	const titleTrimmed = post.title.rendered.trim();
+
+	let excerpt = post.excerpt.rendered;
+	if ( post.excerpt.raw === '' ) {
+		excerpt = post.content.raw;
+	}
+	const excerptElement = document.createElement( 'div' );
+
+	excerptElement.innerHTML = excerpt;
+	excerpt = excerptElement.textContent || excerptElement.innerText || '';
+
+	const dateFormat = __experimentalGetSettings().formats.date; // eslint-disable-line no-restricted-syntax
+
+	return (
+		<div className="wp-block-coblocks-post-carousel__item">
+			{ featuredImageUrl &&
+			<div className="wp-block-coblocks-post-carousel__image">
+				<div className="bg-cover bg-center-center" style={ { backgroundImage: featuredImageStyle } }></div>
+			</div>
+			}
+			<div className={ classnames( 'wp-block-coblocks-post-carousel__content', {
+				'full-height': ! featuredImageUrl,
+			} ) }>
+				{ displayPostDate && post.date_gmt &&
+				<time className="wp-block-coblocks-post-carousel__date" dateTime={ format( 'c', post.date_gmt ) }>
+					{ dateI18n( dateFormat, post.date_gmt ) }
+				</time>
+				}
+				<Disabled>
+					<a alt={ titleTrimmed } href={ post.link } rel="noreferrer noopener" target="_blank">
+						{ titleTrimmed ? (
+							<RawHTML>
+								{ titleTrimmed }
+							</RawHTML>
+						)
+						/* translators: placeholder when a post has no title */
+							: __( '(no title)', 'coblocks' )
+						}
+					</a>
+				</Disabled>
+				{ displayPostContent &&
+				<div className="wp-block-coblocks-post-carousel__excerpt">
+					<RawHTML
+						key="html"
+					>
+						{ excerpt.trim().split( ' ', excerptLength ).join( ' ' ) }
+					</RawHTML>
+				</div>
+				}
+				{ displayPostLink &&
+				<PlainText
+					className="wp-block-coblocks-post-carousel__more-link"
+					onChange={ ( newPostLink ) => setAttributes( { postLink: newPostLink } ) }
+					placeholder={ __( 'Read more', 'coblocks' ) }
+					value={ postLink }
+				/>
+				}
+			</div>
+		</div>
+	);
+};
+
+export default compose( [
+	withSelect( ( select, { clientId } ) => {
+		const blockAttributes = select( 'core/block-editor' ).getBlockAttributes( clientId );
+
+		return {
+			attributes: blockAttributes,
+		};
+	} ),
+] )( PostItem );
```

Comment:
I am confused why this is necessary at all. We are making a store call here when this component has access to the execution context where the `attributes` exist. On line `183` or `159` of `src/blocks/post-carousel/edit.js`. Why not simply pass in the existing attributes? You can modify the `PostItem`s like this:
```JavaScript
<PostItem key={ i } post={ post } { ...{ ...props } } />
```

This change will give you full access to the existing props preventing the need for an extra compose and data-store call.

Reply:
Because the PostItem component is rendered dynamically via the Swiper component (through render-props https://reactjs.org/docs/render-props.html ) , the individual PostItem components are not triggered to re-render via the parent PostEdit. Instead, they are connected to the store and will re-render following any prop update to which they are listening. In the GalleryCarousel a similar pattern is used, though using the Context API.

This way, the Swiper component itself is not re-rendered following any attribute updates, and only the PostItem components are re-rendered specifically in order to minimize any unnecessary re-renders.

Reply:
Minimization of excess renders sounds like a positive for sure. I suspect there is something I am missing still. I guess I am not clear about how if these attributes are already available in the parent and should not change dynamically why we would expect excess renders. When the attribute changes on the parent, the children would re-render as expected because the passed attribute and related useEffect function are watching.

Reply:
I feel like I am about to learn something lol.

Reply:
if you look at the below function within the Swiper component,
```
const renderList = useMemo( () => {
		return list.map( ( item, index ) => (
			<div className={ `swiper-slide` } id={ index } key={ index }>
				{ children( {
					index,
					item,
				} ) }
			</div>
		) );
	}, [ isDraggable ] );
```

The Swiper component is invoked within the PostEdit and uses a single child which is a function : 
```
{ ( { index: i, item: post } ) => {
    return (
	<PostItem clientId={ clientId } key={ i } post={ post } setAttributes={ setAttributes } />
    );
} }
```

The rendering order would be as follows --> PostEdit ---> Swiper --> PostItem. However, within the Swiper component function `renderList`, the function itself is only recalculated following the update of the `isDraggable` variable since the function is memoized. Therefore, the children function would only be updated if that variable is changed. This separates the concern of the Swiper and the child component that it is rendering and allows us to abstractly use the Swiper component without any information of what it is rendering, and when it should update. It would only update when Swiper specific props are changed. That being said, the children component will have the first render of the attributes if we pass it down as props, though when the attribute changes , the swiper component will not update itself with the updated attributes and so instead we can use redux to listen to any store updates , but within the component that needs that information instead of drilling props down through each level.

Reply:
I appreciate this explation of why. This makes sense then to keep.